### DISCUSSION TITLE
Using Telegram infrastracture to distribute packages.

### DISCUSSION THREAD

**Discussion by twaik** [termux/termux-packages#17330] [Ideas]
Hi. Telegram offers unlimited storage so it can be used as cloud storage for termux packages. We can write some bot for this (as a server which keeps references to core files, like `apt/termux-main/dists/stable/Contents-aarch64.gz`) and a bot/script for uploading new files to Telegram servers and removing old/outdated files from there. Also we will need to write something like `apt-transport-https` for this (`apt-transport-mtproto`? `apt-transport-telegram`?). I can help with writing apt transport, but I have no idea how to implement telegram bot.

It will let us save some money on infrastructure.

Answered: https://github.com/termux/termux-packages/discussions/17330#discussioncomment-6342759

**Comment by twaik** [comment]
@termux/package-maintainers

**Comment by ravener** [comment]
I personally don't think that's a viable idea. Sure there may be unlimited storage, so does many other platforms but they were not created for that intention for people to be hosting download servers for non-users of the platform. At worst it may be against their terms of services, even if not it sounds like a lot of work to get a stable infrastructure working with that. That said I don't have experience with Telegram so correct me if I'm wrong but I just don't think this is a good idea.

**Comment by sylirre** [comment]
I have few concerns about this approach:

a. AFAIK Telegram does not allow downloading large media files and especially non-media files without authorization.
b. That's possible violation of terms of service.
c. Even if we implement the apt transport plugin, there still a plain HTTP(S) web server would be required. *Using primary server for this cancels the value of this mirror (if packages.termux.dev is down, this mirror down as well).*
d. Metadata management is much harder.
e. Telegram may reduce the download speed (typically for non-premium users).

Also we have a lot of mirrors nowadays and own primary server. There is no need of switching or looking another place. And I think we shouldn't keep or support "poor man's solutions" that abuse infrastructure of other services. This is not viable in long term and not good for the project in overall.

> It will let us save some money on infrastructure.

We pay around 10 euro per month for packages.termux.dev.

Considering that files bot redirecting to telegram would still require some server. Yes, it could be smaller than current one, but reducing payments by 5 - 7 euros with significant reduction of reliability does not worth it.
